### PR TITLE
Allow multiple RandomIndexedSymbols in Probability Queries

### LINKED ISSUE #20180: Support multivariate expressions in DiscreteMarkovChain
We should be able to process queries of the type 
``` python
>>> T=Matrix([[0.4, 0.3, 0.3],[0.1, 0.4, 0.5],[0.5, 0.25, 0.25]])
>>> Y=DiscreteMarkovChain(Y, [0, 1, 2], T)
>>> P(Eq(Y[1], Y[2]), Eq(Y[3], 1))
>>> P(Gt(Y[1], Y[2]), Eq(Y[3], 1))
```
As discussed in https://github.com/sympy/sympy/pull/20150#issuecomment-700498420

### PR DESCRIPTION
**Pull request by naveensaigit**
I've added support for multiple `RandomIndexedSymbols` in probability queries for `DiscreteMarkovChain`.

### References to other Issues or PRs
Fixes #20180 

#### Release Notes

* stats
    * Added support for multiple RandomIndexedSymbols in DiscreteMarkovChain

### CODE REVIEW

**Comment by naveensaigit** [comment]
@czgdp1807 I added tests and docs for this. Let me know if any changes are required.
I've also corrected the if condition below as this was evaluating to False in some cases
https://github.com/sympy/sympy/blob/a106f4782a9dbe7f8fd16030f15401d977e03ae9/sympy/stats/stochastic_process_types.py#L415-L418
Changed it to
https://github.com/sympy/sympy/blob/8731033bc5139c2c1735ff8407716304c6c3578d/sympy/stats/stochastic_process_types.py#L416-L419

File: sympy/stats/stochastic_process_types.py [hunk 1]
```diff
@@ -413,7 +414,7 @@ def _check_trans_probs(self, trans_probs, row_sum=1):
         if not isinstance(trans_probs, MatrixSymbol):
             rows = trans_probs.tolist()
             for row in rows:
-                if (sum(row) - row_sum) != 0:
+                if abs(sum(row) - row_sum) > 1e-5:
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:417]: L419 suggests to use `Rational`. Probably reviews from more experienced reviewer are needed to decide upon this. @oscarbenjamin @asmeurer @Upabjojr

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:417]: Error pops up despite using `Rational`. This is on Sympy Live
``` python
>>> from sympy import *
>>> from sympy.stats import *
>>> T = Matrix([[Rational(0.5), Rational(0.4), Rational(0.1)], [Rational(0.5), Rational(0.4), Rational(0.1)], [Rational(0.5), Rational(0.4), Rational(0.1)]])
>>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
>>> P(Eq(Y[1], 0), Eq(Y[0], 0))
  File "/base/data/home/apps/s~sympy-live-hrd/66.426491309333028408/sympy/sympy/stats/stochastic_process_types.py", line 280, in _check_trans_probs
    "If you are using Float or floats then please use Rational."%(row_sum))
ValueError: Values in a row must sum to 1. If you are using Float or floats then please use Rational.
```

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:417]:
```python
In [1]: Rational(1, 4)
Out[1]: 1/4

In [2]: Rational(0.4)
Out[2]: 
3602879701896397
────────────────
9007199254740992
```

Comment (Upabjojr) [reply sympy/stats/stochastic_process_types.py:417]: Definitely a bad edit.

Comment (Upabjojr) [reply sympy/stats/stochastic_process_types.py:417]: You shouldn't be using floating-point numbers in SymPy, if possible.

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:417]: My bad, I'll change it back to what it was and change the tests too.

**Comment by naveensaigit** [comment]
The build log says "The job exceeded the maximum time limit for jobs, and has been terminated". I'm not sure what to do here

File: sympy/stats/stochastic_process_types.py [hunk 2]
```diff
@@ -520,20 +521,53 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
...
+                        upper = 1
+                    if isinstance(condition, Gt) or isinstance(condition, Ge):
+                        greater = True
+
+                    for i in range(0, n):
+                        if i <= n//2:
+                            for j in range(0, i + upper):
+                                s += P(Eq(rv[0], i), Eq(rv[1], j)) * P(Eq(rv[1], j), new_given_condition)
+                        else:
+                            s+=P(Eq(rv[0], i), new_given_condition)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:564]:
```suggestion
                            s += self.probability(Eq(rv[0], i), new_given_condition)
```
Calling `P` again is redundant when we know the job will be actually done by `self.probability`.

File: sympy/stats/stochastic_process_types.py [hunk 3]
```diff
@@ -413,7 +414,7 @@ def _check_trans_probs(self, trans_probs, row_sum=1):
         if not isinstance(trans_probs, MatrixSymbol):
             rows = trans_probs.tolist()
             for row in rows:
-                if (sum(row) - row_sum) != 0:
+                if sum(row) - row_sum != 0:
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:417]:
```suggestion
                if (sum(row) - row_sum) != 0:
```
Doesn't really matter much, irrelevant to the motive of this PR.

File: sympy/stats/stochastic_process_types.py [hunk 4]
```diff
@@ -520,20 +521,53 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
...
+                        return Probability(condition)
+            else:
+                min_key_rv = None
                 return Probability(condition)
 
+            if len(rv) > 1:
+                rv = rv[:2]
+                if rv[0].key < rv[1].key:
+                        rv[0], rv[1] = rv[1], rv[0]
+                s = Float(0)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:544]:
```suggestion
                s = Rational(0, 1)
```

File: sympy/stats/stochastic_process_types.py (hunk 2 above)
Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py]: Yes, sorry about that

File: sympy/stats/stochastic_process_types.py [hunk 5]
```diff
@@ -520,20 +520,53 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
...
+
+                    for i in range(0, n):
+                        if i <= n//2:
+                            for j in range(0, i + upper):
+                                s += self.probability(Eq(rv[0], i), Eq(rv[1], j)) * self.probability(Eq(rv[1], j), new_given_condition)
+                        else:
+                            s += self.probability(Eq(rv[0], i), new_given_condition)
+                            for j in range(i + upper, n):
+                                s -= self.probability(Eq(rv[0], i), Eq(rv[1], j)) * self.probability(Eq(rv[1], j), new_given_condition)
+                    return s if greater else 1-s
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:566]:
```suggestion
                    return s if greater else 1 - s
```

File: sympy/stats/stochastic_process_types.py [hunk 6]
```diff
@@ -520,20 +520,53 @@ def probability(self, condition, given_condition=None, evaluate=True, **kwargs):
...
+                rv = rv[:2]
+                if rv[0].key < rv[1].key:
+                        rv[0], rv[1] = rv[1], rv[0]
+                s = Rational(0, 1)
+                n = len(self.state_space)
+
+                if isinstance(condition, Eq) or isinstance(condition, Ne):
+                    for i in range(0, n):
+                        s += self.probability(Eq(rv[0], i), Eq(rv[1], i)) * self.probability(Eq(rv[1], i), new_given_condition)
+                    return s if isinstance(condition, Eq) else 1-s
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:549]:
```suggestion
                    return s if isinstance(condition, Eq) else 1 - s
```

**Comment by czgdp1807** [comment]
Please resolve the conflicts.

**Comment by naveensaigit** [comment]
@czgdp1807 I've resolved all conflicts. This can be merged if no other issues are present

**Comment by czgdp1807** [comment]
LGTM. Will merge if no objections till tomorrow.

**Pull request merged**